primitives/src/block.rs

```diff
@@ -247,6 +241,28 @@ impl fmt::Debug for Header {
     }
 }
 
+impl Encodable for Header {
+    type Encoder = encoding::Encoder6<
+        <Version as Encodable>::Encoder,
+        <BlockHash as Encodable>::Encoder,
+        <TxMerkleNode as Encodable>::Encoder,
+        <BlockTime as Encodable>::Encoder,
+        <CompactTarget as Encodable>::Encoder,
+        encoding::ArrayEncoder<4>,
+    >;
+
+    fn encoder(&self) -> Self::Encoder {
+        encoding::Encoder6::new(
+            self.version.encoder(),
+            self.prev_blockhash.encoder(),
+            self.merkle_root.encoder(),
+            self.time.encoder(),
+            self.bits.encoder(),
+            encoding::ArrayEncoder::without_length_prefix(self.nonce.to_le_bytes()),
```

Comment:
I hacked up a `U32Encoder` and its kinda nice but then it begs the question whether all the 'wrapper' types should have a nested encoder in their encoders or if that is going to far?

```rust
impl Encodable for u32 {
    type Encoder = U32Encoder;
    fn encoder(&self) -> Self::Encoder { U32Encoder::new(*self) }
}

/// An encoder for a `u32` encoded as little-endian bytes.
pub struct U32Encoder(ArrayEncoder<4>);

impl U32Encoder {
    /// Constructs a new `u32` encoder.
    pub fn new(x: u32) -> Self {
        Self(ArrayEncoder::without_length_prefix(x.to_le_bytes()))
    }
}

impl Encoder for U32Encoder {
    fn advance(&mut self) -> Option<&[u8]> {
        self.0.advance()
    }

    fn unadvance(&mut self) { self.0.unadvance() }
}
```

Reply:
I think I've convinced myself we **should** do this. It's a bunch of boilerplate, though this may be a case where macros are acceptable (maybe). If we don't, we're leaking implementation details into our 1.0 API and preventing ourselves from making any meaningful changes to encoder structures.

Reply:
Feeling a little slow here, defining all the low-level encoders makes sense to me just for clarity/organization, but I am not grasping the leaking impl details trade-off. Won't the type of the encoder be exposed no matter how the inner types are encoded?

Reply:
Although this does beg the question whether the `I32Encoder` should be hidden by a `VersionEncoder` also (like in `BlockTimeEncoder`) but not having a different encoder for ever type was discussed early on. Maybe Kix's branch was right and we should have a new type for each?